hummingbot/connector/exchange/kucoin/kucoin_api_order_book_data_source.py

```diff
@@ -145,74 +161,194 @@ async def ws_connect_data(self):
-                            "id": int(time.time()),
-                            "type": "subscribe",
-                            "topic": f"/market/match:{trading_pair}",
-                            "privateChannel": False,
-                            "response": True
-                        }
-                        await ws.send(json.dumps(subscribe_request))
-
-                    async for raw_msg in self._inner_messages(ws):
-                        msg: Dict[str, Any] = json.loads(raw_msg)
-                        if msg["type"] == "pong" or msg["type"] == "ack":
-                            pass
-                        elif msg["type"] == "message":
-                            trading_pair = msg["data"]["symbol"]
-                            data = msg["data"]
-                            trade_message: OrderBookMessage = KucoinOrderBook.trade_message_from_exchange(
-                                data, metadata={"trading_pair": trading_pair}
-                            )
-                            output.put_nowait(trade_message)
-                        else:
-                            self.logger().debug(f"Unrecognized message received from Kucoin websocket: {msg}")
+                if not ws.closed:
+                    await ws.ensure_open()
+                    ping_msg["id"] = int(time.time())
+                    await ws.send(json.dumps(ping_msg))
+            except websockets.exceptions.ConnectionClosedError:
+                pass
             except asyncio.CancelledError:
                 raise
             except Exception:
-                self.logger().error("Unexpected error with WebSocket connection. Retrying after 30 seconds...",
-                                    exc_info=True)
-                await asyncio.sleep(30.0)
+                raise
+            await asyncio.sleep(interval_secs)
 
-    async def listen_for_order_book_diffs(self, ev_loop: asyncio.BaseEventLoop, output: asyncio.Queue):
+    async def get_markets_per_ws_connection(self) -> List[str]:
+        # Fetch the  markets and split per connection
+        all_symbols: List[str] = await self.get_trading_pairs()
+        market_subsets: List[str] = []
+
+        for i in range(0, len(all_symbols), self.SYMBOLS_PER_CONNECTION):
+            symbols_section: List[str] = all_symbols[i: i + self.SYMBOLS_PER_CONNECTION]
+            symbol: str = ','.join(symbols_section)
+            market_subsets.append(symbol)
+
+        return market_subsets
+
+    async def _start_update_tasks(self, stream_type: StreamType, output: asyncio.Queue):
+        self._stop_update_tasks(stream_type)
+        market_assignments = await self.get_markets_per_ws_connection()
+
+        task_dict: dict = {}
+        for task_index, market_subset in enumerate(market_assignments):
+            task_dict[task_index] = {"markets": set(market_subset.split(',')),
+                                     "task": asyncio.ensure_future(self._outer_messages(stream_type,
```

Comment:
All background tasks in hummingbot must use `safe_ensure_future()`, to properly log any background tasks that failed silently.